certbot/certbot/plugins/dns_common_lexicon.py

```diff
@@ -144,3 +161,141 @@ def build_lexicon_config(lexicon_provider_name: str,
+
+    @property
+    @abc.abstractmethod
+    def _provider_name(self) -> str:
+        """
+        The name of the Lexicon provider to use
+        """
+
+    @property
+    def _ttl(self) -> int:
+        """
+        Time to live to apply to the DNS records created by this Authenticator
+        """
+        return 60
+
+    def _add_provider_option(self, creds_var_name: str, creds_var_label: str,
+                             lexicon_provider_option_name: str) -> None:
+        self._provider_options.append(
+            (creds_var_name, creds_var_label, lexicon_provider_option_name))
+
+    def _build_lexicon_config(self, domain: str) -> ConfigResolver:
+        if not hasattr(self, '_credentials'):  # pragma: no cover
+            self._setup_credentials()
+
+        dict_config = {
+            'domain': domain,
+            'provider_name': self._provider_name,
+            'ttl': self._ttl,
+            self._provider_name: {item[2]: self._credentials.conf(item[0])
+                                  for item in self._provider_options}
+        }
+        return ConfigResolver().with_dict(dict_config).with_env()
+
+    def _setup_credentials(self) -> None:
+        self._credentials = self._configure_credentials(
+            key='credentials',
+            label=f'Credentials INI file for {self._provider_name} DNS authenticator',
+            required_variables={item[0]: item[1] for item in self._provider_options},
+        )
+
+    def _perform(self, domain: str, validation_name: str, validation: str) -> None:
+        resolved_domain = self._resolve_domain(domain)
+
+        try:
+            with Client(self._build_lexicon_config(resolved_domain)) as operations:
+                operations.create_record(rtype='TXT', name=validation_name, content=validation)
+        except RequestException as e:
+            logger.debug('Encountered error adding TXT record: %s', e, exc_info=True)
+            raise errors.PluginError('Error adding TXT record: {0}'.format(e))
+
+    def _cleanup(self, domain: str, validation_name: str, validation: str) -> None:
+        try:
+            resolved_domain = self._resolve_domain(domain)
+        except errors.PluginError as e:
+            logger.debug('Encountered error finding domain_id during deletion: %s', e,
+                         exc_info=True)
+            return
+
+        try:
+            with Client(self._build_lexicon_config(resolved_domain)) as operations:
+                operations.delete_record(rtype='TXT', name=validation_name, content=validation)
+        except RequestException as e:
+            logger.debug('Encountered error deleting TXT record: %s', e, exc_info=True)
+
+    def _resolve_domain(self, domain: str) -> str:
+        domain_name_guesses = dns_common.base_domain_name_guesses(domain)
+
+        for domain_name in domain_name_guesses:
+            try:
+                with Client(self._build_lexicon_config(domain_name)):
```

Comment:
I was thinking about the errors people might get here if an older version of Lexicon is used. We can potentially do something fancier if you want now, but I'm not sure it's completely necessary.

```suggestion
                # Using client as a context manager requires `dns-lexicon>=3.14` and we may want to
                # provide better checks and error handling around this in the future.
                with Client(self._build_lexicon_config(domain_name)):
```